oak_functions/loader/src/main.rs

```diff
@@ -43,6 +48,87 @@ pub struct Opt {
+    // TODO(#1930): Support downloading lookup data from a URL instead of a local file.
+    // TODO(#1930): Support periodically re-downloading lookup data at a given time interval.
+    #[structopt(
+        long,
+        help = "Path to a file to load for data lookups, containing key / value entries in protobuf binary format."
+    )]
+    lookup_data_path: String,
+}
+
+struct LookupData {
+    lookup_data_path: String,
+    entries: RwLock<HashMap<Vec<u8>, Vec<u8>>>,
+}
+
+impl LookupData {
+    /// Creates a new [`LookupData`] instance that can refresh its internal entries from the
+    /// provided data file path.
+    ///
+    /// Entries in the data file path must be consecutive binary encoded and length delimited
+    /// protobuf messages according to the definition in `/oak_functions/proto/lookup_data.proto`.
+    ///
+    /// The returned instance is empty, and must be populated by calling the [`LookupData::refresh`]
+    /// method at least once.
+    fn new_empty(lookup_data_path: &str) -> LookupData {
+        LookupData {
+            lookup_data_path: lookup_data_path.to_string(),
+            entries: RwLock::new(HashMap::new()),
+        }
+    }
+
+    /// Refreshes the internal entries of this struct from the data file path provided at
+    /// construction time.
+    ///
+    /// If successful, entries are completely replaced (i.e. not merged).
+    ///
+    /// If there is any error while reading or parsing the data, an error is returned by this
+    /// method, and existing entries are left untouched. The caller may retry the refresh operation
+    /// at a future time.
+    fn refresh(&self) -> anyhow::Result<()> {
+        // TODO(#1930): Avoid loading the entire file in memory for parsing.
+        let lookup_data_buf =
+            fs::read(&self.lookup_data_path).context("could not read lookup data file")?;
+        let entries = load_lookup_entries(&mut lookup_data_buf.as_ref())
+            .context("could not parse lookup data")?;
+        log::info!("loaded {} entries of lookup data", entries.len());
+
+        // This block is here to emphasize and ensure that the write lock is only held for a very
+        // short time.
+        {
+            *self
+                .entries
+                .write()
+                .expect("could not lock entries for write") = entries;
+        }
+
+        Ok(())
+    }
+
+    /// Convenience getter for an individual entry that reduces lock contention by cloning the
+    /// resulting value as quickly as possible and returning it instead of a reference.
+    #[allow(dead_code)]
+    fn get(&self, key: &[u8]) -> Option<Vec<u8>> {
+        self.entries
+            .read()
+            .expect("could not lock entries for read")
+            .get(key)
+            .cloned()
+    }
+}
+
+fn load_lookup_entries<B: bytes::Buf>(
+    lookup_data_buffer: B,
+) -> anyhow::Result<HashMap<Vec<u8>, Vec<u8>>> {
+    let mut lookup_data_buffer = lookup_data_buffer;
```

Comment:
Why is this needed?

Reply:
It's just a rebinding to allow for mutable methods to be invoked on it. The variable is passed by value already, so this is just a detail of the implementation of this method, it's not making things "more" mutable, since the value is already owned by this function.